### ISSUE TITLE
built-in functions bug

### ISSUE THREAD

**Issue by howan36** [open-telemetry/opentelemetry-collector-contrib#36559]
### Component(s)

processor/transform

### Describe the issue you're reporting

expression: - set(attributes["code"], Substring(body,indexOf(body,"code"),indexOf(body,"message")))

error:
      Error: invalid configuration: processors::transform: unable to parse OTTL statement "set(attributes[\"code\"], Substring(body,indexOf(body,\"code\"),indexOf(body,\"message\")))": error while parsing arguments for call to "set": invalid argument at position 1: error while parsing arguments for call to "Substring": invalid argument at position 1: no value field set. This is a bug in the OpenTelemetry Transformation Language

**Comment by bacherfl** [comment]
Hi @howan36!

To make this easier to reproduce, can you please share the full collector config yaml, along with an example input and expected outcome? 
Also, which version of the collector are you using? 

Using the provided example statement, i do get an error as well, however a different one, since the `indexOf` function used in the example is not part of OTTL - when replacing that with int literals, the `Substring` function was parsed correctly

**Comment by howan36** [comment]
transform:
        error_mode: ignore
        log_statements:
# - context: log
          #   conditions:
          #     - IsString(body) == true
          #   statements:
          #     # - set(attributes["trace.id"], Substring(body,body.indexOf("traceId"),body.indexOf("spanId")))
          #     - set(attributes["code"], Substring(body,indexOf(body,"code"),indexOf(body,"message")))

**Comment by howan36** [comment]
i saw ottl support to use this golang built-in function?

**Comment by howan36** [comment]
the int literals you mean like 2,3 numbers? or if you some other method to have the same function

**Comment by bacherfl** [comment]
@howan36  just checked again - `indexOf` does not seem to be a supported function in OTTL. However, can you post an example of a log record body? Then I will try to find an alternative way to extract the code as an attribute

**Comment by howan36** [comment]
response: 
--------------------- request end ---------------------------->
200 http://sssss.xxxxx.svc.cluster.local/v2.0/ciam/xxxx/get/profiles/by/partyUidList

{"code":"0","message":null,"data":[{"customerId":xxxxx,"customerName":null,"gender":0,"phone":"xxxxxx","email":null,"birthYear":null,"birthMonth":null,"birthDay":null,"partyUid":"Cxxxxxxx","customerAddress":null,"wechatOpenId":"oPq0xxxxxxx","wechatUnionid":"ombktxxxxx","customerStatus":"AC","preferStoreName":"xxxxx","preferStoreCode":"549","registerStoreName":"xxxxx","registerStoreCode":"549","registerSource":"{\"content\":\"SHOPPABLE_MINI_P=\"}","ctgProvinceId":null,"ctgProvinceName":null,"ctgCityId":null,"ctgCityName":null,"ctgDistrictId":null,"ctgDistrictName":null,"ctgSubDistrictId":0,"ctgSubDistrictName":null,"postalCode":null,"idCardNumber":null,"registerTime":1730166345903,"childStatus":null,"cdmCustomerChildDtoList":[],"cdmCustomerPreferenceDto":{"decorationPlan":null,"decorationPlanFlag":null,"outboundSmsFlag":true},"ikeaFamilySpaceIdList":[],"cdmMembershipDto":{"cardNumber":"6275980332000698665","wechatCardCode":null,"wechatOfficialAccountOpenId":null},"cardNumber":"6275xxxxxxxx","wechatCardCode":null,"wechatOfficialAccountOpenId":null,"houseOwnerShip":null,"livingStatus":{"status":null,"livingStatusList":[]},"migrationStatus":"N/A"}]}

**Comment by howan36** [comment]
it's unstructed logs, for example, i want to replace the phone field with ****

**Comment by howan36** [comment]
and maybe i will also need to extract its value to attributes

**Comment by bacherfl** [comment]
thanks for the example, I will try to see if there is a way of handling this kind of unstructured log data

**Comment by bacherfl** [comment]
@howan36 maybe something like the following config might be of help to handle unstructured log data like from the example you posted above:

```
extensions:
  health_check:
    endpoint: ${env:MY_POD_IP}:13133
receivers:
  otlp:
    protocols:
      grpc:
        endpoint: 0.0.0.0:4317
processors:
  batch:
    timeout: 10s
  transform:
    log_statements:
      - context: log
        statements:
          - set(attributes["parsedBody"], Split(body, "\n"))
          - set(attributes["parsedBody"], ParseJSON(attributes["parsedBody"][4]))
          - set(attributes["code"], attributes["parsedBody"]["code"])

exporters:
  debug:
    verbosity: detailed

service:
  extensions:
    - health_check
  pipelines:
    logs:
      receivers: [otlp]
      processors:
        - transform
      exporters: [debug]
```

This config first splits the log body lines into a slice, and then takes the line containing the json object and passes that line to the `ParseJSON` function. From there on, properties such as `code` or `phone` can be accessed and modified.
This config does make some assumptions regarding the format of the log entries though, so in this case it relies on the json payload to be in line 4.

**Comment by howan36** [comment]
hi @bacherfl , thanks a lot for your response, so the functions that transform processor supported are all in this page:https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg/ottl/ottlfuncs,  and other go-lang built-in functions won't be supported?

**Comment by bacherfl** [comment]
> hi @bacherfl , thanks a lot for your response, so the functions that transform processor supported are all in this page:https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg/ottl/ottlfuncs, and other go-lang built-in functions won't be supported?

Exactly, the functions listed in those docs are the ones that are supported. Executing Golang functions directly is not possible

**Comment by evan-bradley** [comment]
@howan36 Would the [ExtractPatterns](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/pkg/ottl/ottlfuncs#extractpatterns) function work for your use case?
